cpp/src/sqls/sqls_rtti_comp.h

```diff
@@ -127,198 +260,173 @@ struct LesserRTTI
     return (static_cast<const ColType*>(columns[col_index]))[row];
   }
   
+  __device__
+  static bool is_valid(int col_index,
+                       IndexT row,
+                       const gdf_valid_type* const * valids)
+  {
+    return gdf_is_valid(valids[col_index], row);
+  }
+
 private:
   enum class State {False = 0, True = 1, Undecided = 2};
 
   struct OpLess
   {
-    __device__
-    OpLess(IndexT row1, IndexT row2):
-      row1_(row1),
-      row2_(row2)
-    {
-    }
-    
     template<typename ColType>
     __device__
-    State operator() (int col_index,
-                      const void* const * columns,
-                      ColType)
+    State operator() (IndexT row1, IndexT row2,
+                      int col_index,
+                      const void* const * columns)
     {
-      ColType res1 = LesserRTTI::at<ColType>(col_index, row1_, columns);
-      ColType res2 = LesserRTTI::at<ColType>(col_index, row2_, columns);
+      ColType res1 = LesserRTTI::at<ColType>(col_index, row1, columns);
+      ColType res2 = LesserRTTI::at<ColType>(col_index, row2, columns);
       
       if( res1 < res2 )
         return State::True;
       else if( res1 == res2 )
         return State::Undecided;
       else
-	return State::False;
+	      return State::False;
     }
-  private:
-    IndexT row1_;
-    IndexT row2_;
   };
 
-  struct OpEqual
+  struct OpLess_with_nulls
   {
+    template<typename ColType>
+	  __device__
+	  State operator() (IndexT row1, IndexT row2,
+                      int col_index,
+                      const void* const * columns,
+                      const gdf_valid_type* const * valids,
+                      bool nulls_are_smallest)
+	  {
+		  ColType res1 = LesserRTTI::at<ColType>(col_index, row1, columns);

```

Comment:
I think this logic could be both be made more efficient and easier to understand with something like this: 

```
      auto initializer = [nulls_are_smallest, col_index, columns, valids](IndexT row)
      {
        if(LesserRTTI::is_valid(col_index, row, valids))
          return LesserRTTI::at<ColType>(col_index, row, columns);
        
        if(nulls_are_smallest)
          return std::numeric_limits<ColType>::lowest();
        
        return std::numeric_limits<ColType>::max();
      }

      ColType const res1 = initializer(row1);
      ColType const res2 = initializer(row2);

      if( res1 < res2 )
        return State::True;
      else if( res1 == res2 )
        return State::Undecided;
      else
        return State::False;
```

Likewise applies to `OpGreater_with_nulls`

Reply:
Nevermind, this won't work. 

If you had NULL < 255, and `nulls_are_smallest = false`, you would expect `NULL < 255` to return `false`. But by my logic, you would get `undecided` because it would be `255 == 255